File: src/java/shared/src/main/java/com/aurora/shared/exception/ExceptionInterceptor.java
package com.aurora.shared.exception;

import io.grpc.*;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

/**
 * gRPC Server Interceptor: Maps Java domain exceptions sang gRPC status codes.
 * Parallel với ExceptionInterceptor.cs trong .NET.
 */
public class ExceptionInterceptor implements ServerInterceptor {

    private static final Logger log = LoggerFactory.getLogger(ExceptionInterceptor.class);

    @Override
    public <ReqT, RespT> ServerCall.Listener<ReqT> interceptCall(
            ServerCall<ReqT, RespT> call,
            Metadata headers,
            ServerCallHandler<ReqT, RespT> next) {

        ServerCall.Listener<ReqT> listener = next.startCall(call, headers);

        return new ForwardingServerCallListener.SimpleForwardingServerCallListener<ReqT>(listener) {
            @Override
            public void onHalfClose() {
                try {
                    super.onHalfClose();
                } catch (Throwable t) {
                    handleException(call, t);
                }
            }
        };
    }

    private <ReqT, RespT> void handleException(ServerCall<ReqT, RespT> call, Throwable t) {
        if (t instanceof DomainExceptions.NotFoundException ex) {
            call.close(Status.NOT_FOUND.withDescription(ex.getMessage()), new Metadata());
        } else if (t instanceof DomainExceptions.ConflictException ex) {
            call.close(Status.ALREADY_EXISTS.withDescription(ex.getMessage()), new Metadata());
        } else if (t instanceof DomainExceptions.ForbiddenException ex) {
            call.close(Status.PERMISSION_DENIED.withDescription(ex.getMessage()), new Metadata());
        } else if (t instanceof DomainExceptions.DomainException ex) {
            call.close(Status.INVALID_ARGUMENT.withDescription(ex.getMessage()), new Metadata());
        } else if (t instanceof StatusRuntimeException ex) {
            call.close(ex.getStatus(), ex.getTrailers());
        } else {
            log.error("Unhandled exception in gRPC handler: {}", t.getMessage(), t);
            call.close(Status.INTERNAL.withDescription("Internal server error"), new Metadata());
        }
    }
}

File: src/dotnet/shared/Interceptors/ExceptionInterceptor.cs
using Grpc.Core;
using Grpc.Core.Interceptors;
using Microsoft.Extensions.Logging;
using Shared.Exceptions;

namespace Shared.Interceptors;

/// <summary>
/// Map domain exceptions sang gRPC status codes.
/// Exception không xác định → Internal (không leak chi tiết ra ngoài, chỉ ghi log).
/// </summary>
public class ExceptionInterceptor(ILogger<ExceptionInterceptor> logger) : Interceptor
{
    public override async Task<TResponse> UnaryServerHandler<TRequest, TResponse>(
        TRequest request,
        ServerCallContext context,
        UnaryServerMethod<TRequest, TResponse> continuation)
    {
        try
        {
            return await continuation(request, context);
        }
        catch (NotFoundException ex)
        {
            throw new RpcException(new Status(StatusCode.NotFound, ex.Message));
        }
        catch (ConflictException ex)
        {
            throw new RpcException(new Status(StatusCode.AlreadyExists, ex.Message));
        }
        catch (DomainException ex)
        {
            throw new RpcException(new Status(StatusCode.InvalidArgument, ex.Message));
        }
        catch (ForbiddenException ex)
        {
            throw new RpcException(new Status(StatusCode.PermissionDenied, ex.Message));
        }
        catch (RpcException)
        {
            throw; // đã là gRPC status hợp lệ — không wrap lại
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "Unhandled exception in gRPC method {Method}", context.Method);
            throw new RpcException(new Status(StatusCode.Internal, "Internal server error"));
        }
    }

    public override async Task ServerStreamingServerHandler<TRequest, TResponse>(
        TRequest request,
        IServerStreamWriter<TResponse> responseStream,
        ServerCallContext context,
        ServerStreamingServerMethod<TRequest, TResponse> continuation)
    {
        try
        {
            await continuation(request, responseStream, context);
        }
        catch (NotFoundException ex)
        {
            throw new RpcException(new Status(StatusCode.NotFound, ex.Message));
        }
        catch (ConflictException ex)
        {
            throw new RpcException(new Status(StatusCode.AlreadyExists, ex.Message));
        }
        catch (DomainException ex)
        {
            throw new RpcException(new Status(StatusCode.InvalidArgument, ex.Message));
        }
        catch (ForbiddenException ex)
        {
            throw new RpcException(new Status(StatusCode.PermissionDenied, ex.Message));
        }
        catch (RpcException)
        {
            throw;
        }
        catch (Exception ex)
        {
            logger.LogError(ex, "Unhandled exception in gRPC method {Method}", context.Method);
            throw new RpcException(new Status(StatusCode.Internal, "Internal server error"));
        }
    }
}
